File: rust/codegen/src/naming.rs
//! Port of `rust/project/naming.rb` — read that file's own comments for
//! the full rationale behind each rule; this mirrors its algorithm
//! directly, function for function.

pub fn scalar_rust_type(type_name: &str) -> Option<&'static str> {
    match type_name {
        "String" => Some("String"),
        "Integer" => Some("i64"),
        "Float" => Some("f64"),
        _ => None,
    }
}

pub fn reference_type(type_name: &str) -> bool {
    type_name.starts_with("Reference<")
}

/// `X` out of `Reference<X>` — `None` for anything that isn't a reference
/// type at all.
pub fn reference_target(type_name: &str) -> Option<&str> {
    type_name.strip_prefix("Reference<").and_then(|rest| rest.strip_suffix('>'))
}

pub fn effective_scalar_type(type_name: &str) -> Option<&'static str> {
    if reference_type(type_name) {
        return Some("String");
    }
    match type_name {
        "String" => Some("String"),
        "Integer" => Some("Integer"),
        "Float" => Some("Float"),
        _ => None,
    }
}

pub fn rust_type(type_name: &str, list: bool) -> String {
    let scalar = effective_scalar_type(type_name);
    let inner = match scalar {
        Some(s) => scalar_rust_type(s).unwrap().to_string(),
        None => rust_ident(type_name),
    };
    if list {
        format!("Vec<{inner}>")
    } else {
        inner
    }
}

pub fn rust_ident(name: &str) -> String {
    name.chars().filter(|c| c.is_ascii_alphanumeric()).collect()
}

/// `cmd.gsub(/(?<=.)([A-Z])/, '_\1').downcase` — insert `_` before every
/// uppercase letter that isn't the very first character, then lowercase.
pub fn dispatch_fn_name(cmd: &str) -> String {
    let mut out = String::new();
    for (i, c) in cmd.chars().enumerate() {
        if i > 0 && c.is_ascii_uppercase() {
            out.push('_');
        }
        out.push(c);
    }
    out.to_lowercase()
}

pub const RUST_KEYWORDS: &[&str] = &[
    "as", "break", "const", "continue", "crate", "dyn", "else", "enum", "extern", "false", "fn", "for", "if", "impl", "in", "let", "loop", "match", "mod", "move", "mut", "pub", "ref",
    "return", "self", "Self", "static", "struct", "super", "trait", "true", "type", "unsafe", "use", "where", "while", "abstract", "become", "box", "do", "final", "macro", "override", "priv",
    "typeof", "unsized", "virtual", "yield", "try",
];

pub fn rust_field(name: &str) -> String {
    name.to_string()
}

pub fn rust_ident_field(name: &str) -> String {
    let field = rust_field(name);
    if RUST_KEYWORDS.contains(&field.as_str()) {
        format!("r#{field}")
    } else {
        field
    }
}

/// A closed-set member's value is business text, not a pre-sanitized Rust
/// identifier — split on ANY run of non-alphanumeric characters (see
/// naming.rb's own header on why plain `_`/whitespace splitting broke on
/// glob-shaped members like `"*.port"`).
pub fn closed_set_variant(value: &str) -> String {
    value
        .split(|c: char| !c.is_ascii_alphanumeric())
        .filter(|s| !s.is_empty())
        .map(capitalize)
        .collect::<Vec<_>>()
        .join("")
}

fn capitalize(word: &str) -> String {
    let mut chars = word.chars();
    match chars.next() {
        None => String::new(),
        Some(first) => first.to_uppercase().collect::<String>() + &chars.as_str().to_lowercase(),
    }
}

/// `name.to_s.gsub(/([a-z0-9])([A-Z])/, '\1_\2').upcase`
pub fn screaming_snake(name: &str) -> String {
    let chars: Vec<char> = name.chars().collect();
    let mut out = String::new();
    for (i, &c) in chars.iter().enumerate() {
        if i > 0 {
            let prev = chars[i - 1];
            if (prev.is_ascii_lowercase() || prev.is_ascii_digit()) && c.is_ascii_uppercase() {
                out.push('_');
            }
        }
        out.push(c);
    }
    out.to_uppercase()
}

pub fn scalar_to_value(type_name: &str, rust_expr: &str) -> Option<String> {
    match type_name {
        "String" => Some(format!("Value::Str({rust_expr}.clone())")),
        "Integer" => Some(format!("Value::Int({rust_expr})")),
        "Float" => Some(format!("Value::Float({rust_expr})")),
        _ => None,
    }
}

/// A literal mutation-source RHS — mirrors `naming.rb#literal_rhs`. Takes
/// the already-parsed `crate::ruby_value::Value`-shaped JSON literal
/// (String/Integer/Float/Bool) this crate reads out of `ir.json`.
pub fn literal_rhs(literal: &crate::json::Json) -> String {
    match literal {
        crate::json::Json::String(s) => format!("{}.to_string()", ruby_inspect_string(s)),
        // `Integer, Float then literal.to_s` — ONE Ruby case, but `to_s`
        // renders differently per real type (`0.to_s == "0"`, `0.0.to_s ==
        // "0.0"`) — kept as two Rust match arms so a whole-number Float
        // default still renders as a valid `f64` literal (`0.0`, not the
        // integer-typed `0` that would fail to unify with the `x.as_f64()`
        // branch it sits beside in `scalar_from_json_expr`'s own output).
        crate::json::Json::Int(n) => n.to_string(),
        crate::json::Json::Float(_) => literal.to_s(),
        crate::json::Json::Bool(b) => b.to_string(),
        other => panic!("unsupported literal mutation source {other:?} — not one of String/Integer/Float/Boolean"),
    }
}

/// Ruby's `String#inspect` — a double-quoted, backslash/quote-escaped
/// literal. Shared by every codegen site that needs to embed a Ruby-
/// literal-shaped string constant in generated Rust text (matching Rust's
/// own escaping for the common cases both languages agree on).
pub fn ruby_inspect_string(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 2);
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\t' => out.push_str("\\t"),
            '\r' => out.push_str("\\r"),
            _ => out.push(c),
        }
    }
    out.push('"');
    out
}

File: rust/project/naming.rb
module RustProjection
  module Projector
    module_function

    SCALAR      = { "String" => "String", "Integer" => "i64", "Float" => "f64" }.freeze
    SCALAR_KIND = { "String" => :string, "Integer" => :int, "Float" => :float }.freeze

    # `Reference<X>` is not a scalar per the IR's own vocabulary, but it
    # behaves like one for codegen purposes: aggregates-and-value-objects.md's
    # "Pointing at another aggregate" section is explicit that a reference
    # "is a bare id — a String — not a nested object." So a reference is
    # represented as a plain `String`, the same as any other id, at the
    # STRUCT-FIELD level. `resolve_references` (Ruby's own existence check —
    # `command_rules/references.rb`, read directly) is now generated too,
    # but at the registry level (`reactions.rb`'s `reference_check`,
    # `registry.rb`'s own emit), which is where Ruby's own version reaches
    # through `@registry.repository(domain, target)` — not here, and not by
    # changing this field's Rust type.
    def reference_type?(type_name) = type_name.to_s.start_with?("Reference<")

    # `X` out of `Reference<X>` — the target aggregate's own bare name, the
    # same string `command_rules/references.rb#referenced_aggregate` reaches
    # via `attribute.type.resolve.name`. `nil` for anything that isn't a
    # reference type at all, same shape as `reference_type?`.
    def reference_target(type_name)
      match = type_name.to_s.match(/\AReference<(.+)>\z/)
      match && match[1]
    end

    def effective_scalar_type(type_name)
      return "String" if reference_type?(type_name)

      type_name if SCALAR.key?(type_name)
    end

    def rust_type(type_name, list:)
      scalar = effective_scalar_type(type_name)
      inner = scalar ? SCALAR.fetch(scalar) : rust_ident(type_name)
      list ? "Vec<#{inner}>" : inner
    end

    def rust_ident(name) = name.to_s.gsub(/[^A-Za-z0-9]/, "")
    def dispatch_fn_name(cmd) = cmd.gsub(/(?<=.)([A-Z])/, '_\1').downcase

    # Two different jobs a field name does in generated code, and they must
    # NOT share one function: as a STRING LITERAL (a `Fielded::field` match
    # arm's key, a payload map key) the plain name is correct Rust — `"type"`
    # IS the string a caller passes to look a field up by name. As an
    # IDENTIFIER (a struct field, `self.X`/`args.X`/`record.X`) a name that
    # collides with a Rust keyword is a syntax error unless raw-escaped —
    # found live: the self-hosted grammar's own `Attribute`/`Argument`
    # aggregates declare a field literally named `type`, because that is
    # what it is.
    RUST_KEYWORDS = %w[
      as break const continue crate dyn else enum extern false fn for if impl
      in let loop match mod move mut pub ref return self Self static struct
      super trait true type unsafe use where while abstract become box do
      final macro override priv typeof unsized virtual yield try
    ].freeze

    def rust_field(name) = name.to_s
    def rust_ident_field(name)
      field = rust_field(name)
      RUST_KEYWORDS.include?(field) ? "r##{field}" : field
    end

    # A closed-set member's value is business text, not a pre-sanitized Rust
    # identifier — splitting on `_`/whitespace alone was enough for every
    # value pizzas/banking ever declared ("small", "large", "open"...), but
    # the self-hosted grammar's own Vocabulary chapter has closed sets whose
    # members are glob patterns ("*.port", "Translations/*.bluebook") —
    # `.capitalize` leaves `*`/`.`/`/` untouched, so the OLD version emitted
    # `Translations/*.bluebook` as a literal enum variant line, and Rust's
    # lexer read the embedded `/*` as an unterminated block comment. Split on
    # ANY run of non-alphanumeric characters instead, so every character that
    # isn't a valid identifier constituent is a word boundary, not preserved
    # text.
    def closed_set_variant(row)
      _, value = row.first
      value.to_s.split(/[^A-Za-z0-9]+/).reject(&:empty?).map(&:capitalize).join
    end

    def screaming_snake(name)
      name.to_s.gsub(/([a-z0-9])([A-Z])/, '\1_\2').upcase
    end

    def scalar_to_value(type_name, rust_expr)
      case type_name
      when "String"  then "Value::Str(#{rust_expr}.clone())"
      when "Integer" then "Value::Int(#{rust_expr})"
      when "Float"   then "Value::Float(#{rust_expr})"
      end
    end

    # TRUE for anything `fielded.rb`'s three "is this attribute a nested
    # value object" sites can safely emit `Field::Nested(&self.x)` for —
    # every ordinary (non-closed-set) VO always could; a single-field
    # CLOSED SET (an enum, `emit_closed_set_enum`) now can too, once it
    # carries its own `Fielded` impl (`emit_closed_set_fielded_impl`,
    # below) answering "value" the same way any other single-field VO's
    # own generated `Fielded` impl already does. FALSE for a MULTI-FIELD
    # closed set (`emit_closed_set_table` — Syntax::Keyword, Argument):
    # genuinely no `Fielded` impl exists or is proven needed for that
    # shape (json_codec.rb's own `emit_closed_set_codec` header: "nothing
    # in either example domain looks one up generically or passes one as
    # a command argument") — `Field::Nested` would fail to compile
    # against it, and nothing in the corpus has ever needed it to.
    #
    # FOUND LIVE, a real, previously-invisible gap: no domain ever
    # declared a `given`/expression referencing a closed-set-typed
    # attribute before lifeadelics' vendored embryonaut_bluebooks/
    # payments (`Payment::Succeed`'s own "the processor matches..."
    # given, over `processor: Processor` and `reported_processor:
    # Processor`, both `one_of:` closed sets) — found live, a reaction-
    # triggered Payment::Succeed refusing with "cannot resolve
    # \"processor\" — no such attribute or argument" even though the
    # aggregate held a real, rehydrated `processor` value the whole time.
    def fielded_capable_nested?(vo)
      !vo[:closed_set] || vo[:attributes].size == 1
    end

    # THE `Fielded` IMPL A SINGLE-FIELD CLOSED SET NEVER GOT —
    # `emit_value_object`'s own single-field-closed-set branch (types.rb)
    # used to return early with JUST the bare enum (`emit_closed_set_
    # enum`), matching neither `emit_fielded_flat`'s shape (it has no
    # `attributes` to iterate, being a tag, not a struct) nor needing one
    # before `fielded_capable_nested?` above started asking. Answers
    # "value" — the SAME single field every other closed-set codec
    # already treats it as (`emit_closed_set_codec`'s own `to_json`:
    # `Json::obj(vec![("value", ...)])`) — with the identical match this
    # enum's own `to_json` already builds, reproduced here rather than
    # shared via a common helper method (lower-risk: purely additive,
    # touches nothing about the enum's own EXISTING to_json/from_json
    # generation or any other caller of it).
    def emit_closed_set_fielded_impl(vo)
      name = rust_ident(vo[:name])
      arms = vo[:members].map do |row|
        _, raw = row.first
        "#{name}::#{closed_set_variant(row)} => #{raw.to_s.inspect}.to_string(),"
      end.join(" ")

      "impl crate::kernel::Fielded for #{name} {\n" \
        "    fn field(&self, name: &str) -> Option<crate::kernel::Field<'_>> {\n" \
        "        use crate::kernel::{Field, Value};\n" \
        "        match name {\n" \
        "            \"value\" => Some(Field::Value(Value::Str(match self { #{arms} }))),\n" \
        "            _ => None,\n" \
        "        }\n" \
        "    }\n" \
        "}"
    end

    def literal_rhs(literal)
      case literal
      when String        then "#{literal.inspect}.to_string()"
      when Integer, Float then literal.to_s
      when true, false     then literal.to_s
      else raise "unsupported literal mutation source #{literal.inspect} — not one of String/Integer/Float/Boolean"
      end
    end
  end
end
